File: 송하연/4주차/두큐합같게만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = -2;
        
        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();
        
        int sum1 = 0;
        int sum2 = 0;
        
        for(int i : queue1){
            q1.offer(i); 
            sum1 += i;
        }
        for(int i : queue2){
            q2.offer(i); 
            sum2 += i ;
        }

        if((sum1+sum2)%2 == 1){return -1;}
        int target = (sum1+sum2)/2;
        
        
        int cnt = 0;
        int maxCount = (queue1.length + queue2.length)*2;
        
        
        while((sum1 != target || sum2 != target) && cnt < maxCount){
    
            if(sum1 > target){
                if(q1.isEmpty()) return -1; 
                int num = q1.poll();
                q2.offer(num);
                
                sum1 -= num;
                sum2 += num;
                
                cnt++;
    
            }
            else if (sum2 > target){
                if(q2.isEmpty()) return -1; 
                int num = q2.poll();
                q1.offer(num);
                
                sum1 += num;
                sum2 -= num;
                
                cnt++;
            }
          
        }

        if(cnt >= maxCount){return -1;}
        else {answer =  cnt;}
        
        return answer;
    }
    
}

File: 이시우/4주차/두 큐 합 같게 만들기.cpp
#include <string>
#include <vector>
#include <numeric>
#include <queue>
#include <iostream>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    long long sumA = accumulate(queue1.begin(), queue1.end(), 0);
    long long sumB = accumulate(queue2.begin(), queue2.end(), 0);
    
    // 합이 홀수인 경우
    if((sumA + sumB) % 2 != 0) return -1;
    
    queue<int> q1;
    queue<int> q2;
    
    for(const int &i : queue1) q1.push(i);
    for(const int &i : queue2) q2.push(i);

    
    int answer = 0;
    while (sumA != sumB) {
        // 큐가 비었을 경우 불가능
        // 한바퀴 이상 돌았을 경우 불가능
        if(!q1.size() || !q2.size() || answer > (q1.size()+q2.size())+1){
            answer = -1;
            break;
        }
        
        // A가 더 클때
        if(sumA > sumB){
            int v = q1.front();
            q1.pop();
            q2.push(v);
            sumA -= v;
            sumB += v;
            answer += 1;
        } else {
            int v = q2.front();
            q2.pop();
            q1.push(v);
            sumA += v;
            sumB -= v;
            answer += 1;
        }
    }
    
    return answer;
}
